discovery.md

```diff
@@ -0,0 +1,436 @@
+structure describing the CloudEvent.
+
+#### Intermediary
+
+An "intermediary" receives a message containing an event for the purpose of
+forwarding it to the next receiver, which might be another intermediary or a
+consumer. A typical task for an intermediary is to route the event to receivers
+based on the information in the event context.
+
+#### Source
+
+The "source" is the context in which the occurrence happened. In a distributed
+system it might consist of multiple Producers. If a source is not aware of
+CloudEvents, an external producer creates the CloudEvent on behalf of the
+source.
+
+#### Consumer
+
+A "consumer" receives the event and acts upon it. It uses the context and data
+to execute some logic, which might lead to the occurrence of new events.
+
+#### Subscription
+
+The request for events from an Event Producer system.
+
+#### Event Subscriber
+
+The entity managing the lifecycle of a Subscription on behalf of an Event
+Consumer. In some instances this might be the same entity as the Event Consumer.
+
+## API Specification
+
+The discovery API takes the declarative API design approach. This document
+structure focuses on discovery where each document has a two part key of
+`producer` and `type`. This takes a denormalized view where the composite key
+of documents is the `producer` and `type`. Producer is defined as an arbitrary
+human readable string and type is the CloudEvent type attribute.
+
+### Producer entity Attributes
+
+The `Producer` entity is the main component of the discovery API. This section
+covers the structure of that entity and the next section describes the
+requirements for the query API.
+
+#### producer
+
+- Type: `String`
+- Description: Identifies the producer of the event. The producer string SHOULD
+  be human readable and can identify a top level producer system.
+- Constraints:
+  - REQUIRED
+  - MUST be a non-empty string
+- Examples:
+  - "GitHub.com"
+  - "Awesome Cloud Storage"
+  - "com.example.microservices.userlogin"
+
+#### produceruri
+
+- Type: `URI`
+- Description: Absolute URI that provides a link back to the producer, or
+  documentation about the producer. This is for human consumption.
+- Constraints:
+  - OPTIONAL
+  - If present, MUST be a non-empty URI
+- Examples:
+  - "http://cloud.example.com/docs/blobstorage"
+
+#### source
+
+- Type: `List` of `String`
+- Description: CloudEvents [`source`](https://github.com/cloudevents/spec/blob/master/spec.md#source-1)
+  attribute values available from this producer for this event type. Values here
+  may be used to create subscriptions from this producer.
+- Constraints
+  - OPTIONAL
+- Examples:
+  - Single Internet-wide unique URI with a DNS authority.
+   - ["https://github.com/cloudevents"]
+  - Application-specific identifiers
+   - ["/sensors/tn-1234567/alerts", "/sensors/tr-1234567/alerts"]
+
+#### sourcestructure
+
+- Type: `String`
+- Description: Documentation of the format of the source attribute for this
```

Comment:
Is the intent that we define the template language used?

Reply:
Good question for the group.

That wasn't my intention as the items in there can't necessarily be interpreted by tooling.

Reply:
Ah, in that case we may want to say that this is for human consumption

Reply:
ok, based on some feedback I got - I removed this for this cut.